File: src/main/java/org/joel3112/componentbuilder/ui/settingsComponent/components/ItemsComponent.java
package org.joel3112.componentbuilder.ui.settingsComponent.components;

import com.intellij.ui.components.JBTextField;
import com.intellij.util.ui.FormBuilder;
import org.joel3112.componentbuilder.settings.data.Item;
import org.joel3112.componentbuilder.settings.data.MutableState;
import org.joel3112.componentbuilder.ui.settingsComponent.Component;

import javax.swing.*;
import java.util.ArrayList;
import java.util.List;

public class ItemsComponent implements Component, MutableState<List<Item>> {
    private final List<JBTextField> nameTextFields;

    public ItemsComponent(List<Item> items) {
        nameTextFields = new ArrayList<>();

        for (Item item : items) {
            nameTextFields.add(new JBTextField(item.getName()));
        }
    }

    @Override
    public void addToBuilder(FormBuilder formBuilder) {
        JPanel panel = new JPanel();
        for (int i = 0; i < nameTextFields.size(); i++) {
            JPanel itemPanel = new JPanel();
            itemPanel.add(new JLabel("Item:" + (i + 1)));
            itemPanel.add(nameTextFields.get(i));
            panel.add(itemPanel);
        }
        formBuilder.addLabeledComponent(panel, new JLabel());
    }

    @Override
    public List<Item> getState() {
        List<Item> values = new ArrayList<>();
        for (int i = 0; i < nameTextFields.size(); i++) {
            values.add(new Item(nameTextFields.get(i).getText()));
        }
        return values;
    }

    @Override
    public void setState(List<Item> items) {
        for (int i = 0; i < items.size(); i++) {
            nameTextFields.get(i).setText(items.get(i).getName());
        }
    }
}

File: src/main/kotlin/org/joel3112/componentbuilder/settings/ui/settingsComponent/components/ItemsComponent.kt
package org.joel3112.componentbuilder.settings.ui.settingsComponent.components

import com.intellij.ui.components.JBTextField
import com.intellij.util.ui.FormBuilder
import org.joel3112.componentbuilder.settings.data.Item
import org.joel3112.componentbuilder.settings.data.MutableState
import org.joel3112.componentbuilder.settings.ui.settingsComponent.Component
import javax.swing.JLabel
import javax.swing.JPanel

class ItemsComponent(items: List<Item>) : Component, MutableState<List<Item>> {
    private val nameTextFields: MutableList<JBTextField> = ArrayList()

    init {
        for ((name) in items) {
            nameTextFields.add(JBTextField(name))
        }
    }

    override fun addToBuilder(formBuilder: FormBuilder) {
        val panel = JPanel()
        for (i in nameTextFields.indices) {
            val itemPanel = JPanel()
            itemPanel.add(JLabel("Item:" + (i + 1)))
            itemPanel.add(nameTextFields[i])
            panel.add(itemPanel)
        }
        formBuilder.addLabeledComponent(panel, JLabel())
    }

    override var state: List<Item>
        get() {
            val values: MutableList<Item> = ArrayList()
            for (i in nameTextFields.indices) {
                values.add(Item(nameTextFields[i].text))
            }
            return values
        }
        set(items) {
            for (i in items.indices) {
                nameTextFields[i].text = items[i].name
            }
        }
}
